src/Accelerators/NNPA/Conversion/ONNXToZHigh/RewriteONNXForZHigh.cpp

```diff
@@ -415,9 +420,165 @@ class SplitLargeMatMulPattern : public OpRewritePattern<ONNXMatMulOp> {
   }
 };
 
+class ExpandAddConstantPattern : public OpRewritePattern<ONNXAddOp> {
+public:
+  using OpRewritePattern<ONNXAddOp>::OpRewritePattern;
+
+  /*
+    If add does not have a matmul input, check constant to be uni-broadcastable
+    to the non-constant term.
+
+    If add has a matmul input, check for these patterns.
+
+    Arch 14
+    type: input_a	input_b	input_c -> result
+    unstacked: ZDNN_2D(m,n) ZDNN_2D(n,p) ZDNN_1D(p) -> ZDNN_2D(m,p)
+    stacked: ZDNN_3DS(s,m,n) ZDNN_3DS(s,n,p) ZDNN_2DS(p) -> ZDNN_3DS(s,m,p)
+      => ZDNN_3DS(s,m,n) ZDNN_3DS(s,n,p) ZDNN_2DS(s, 1, p) -> ZDNN_3DS(s,m,p)
+
+    Arch 15: Arch 14 plus patterns below,
+    bcast1: ZDNN_2D(m,n) ZDNN_3DS(s,n,p) ZDNN_2DS(p) -> ZDNN_3DS(s,m,p)
+      => ZDNN_2D(m,n) ZDNN_3DS(s,n,p) ZDNN_2DS(s,1,p) -> ZDNN_3DS(s,m,p)
+
+    bcast23: ZDNN_3DS(s,m,n) ZDNN_2D(n,p) ZDNN_1D(p) -> ZDNN_3DS(s,m,p)
+
+    staked and bcast1: Hardware need 2DS(s, p). To make broadcast legal, have to
+    add a "1" in between the s and the p. If we have only a (p), it works too
+    when we expand the added constant to s replicas of (p) to become (s, 1 p).
+    Hardware op will eventually get (s, p).
+  */
+  LogicalResult matchAndRewrite(
+      ONNXAddOp addOp, PatternRewriter &rewriter) const override {
+    // For the moment, only handle constant shapes.
+    // TODO: add support for dynamic shapes that do not impact the constants.
+    if (!hasStaticShape(addOp.getA().getType()) ||
+        !hasStaticShape(addOp.getB().getType()))
```

Comment:
it seems only B has static shape would be enough since we only need dimensions `s` and `p` to expand the input C from shape `(p)` to `(s, p)`. And B is often a constant.

Reply:
Yes, changed to only checking if B is constant. Had to modify the code slightly to accommodate this, it now works.